### PR TITLE
fix(data structures): default committee size after transitioning from…

### PR DESCRIPTION
**Pull request by girazoki**
This PR fixes the transition from bootstrap committee to ARS based committee. Essentially, as soon as the epoch limit is reached and as long as the boootstrap committee keeps being the last committee, returns the default committee size. I implemented like this because the committee size is always calculated based on the number of votes we had in the previous superblock, which is the bootstrap committee length at the moment of transitioning.

In short, not very convinced about this

 However, this has some caveats:

- If the ARS members are offline at the moment of transitioning, the committee size does not get reduced.
- If after transitioning, for some reason the ARS is identical to the bootstrap committee, and one of them goes offline, the committee size does not get reduced

### CODE REVIEW

File: node/src/actors/chain_manager/mod.rs [hunk 1]
```diff
@@ -1251,46 +1251,53 @@ impl ChainManager {
...
+                                    let reputed = reputed_ars(&ars_list, &reputation_engine);
+                                    // In case of no reputed nodes, return all active nodes
+                                    if reputed.is_empty() {
+                                        ars_list
+                                    } else {
+                                        reputed
+                                    }
+                                };
+                                let committee_size = {
+                                    if act.chain_state.superblock_state.get_committee() == bootstrap_committee.into_iter().collect::<HashSet<PublicKeyHash>>() {
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mod.rs:1275]: This checks if the last superblock has been signed by the bootstrap committee. But wouldn't it be easier to just check the superblock index, and compare it to a variable like `last_superblock_index_that_will_be_signed_by_the_bootstrap_committee`?

File: data_structures/src/superblock.rs [hunk 2]
```diff
@@ -438,6 +438,11 @@ impl SuperBlockState {
         self.votes_mempool.get_received_votes()
     }
 
+    /// Returns the current committee
+    pub fn get_committee(&self) -> HashSet<PublicKeyHash> {
+        self.signing_committee.clone()
```
Comment (tmpolaczyk) [inline data_structures/src/superblock.rs:443]: You should be able to remove this clone and use a reference instead

```suggestion
    pub fn get_committee(&self) -> &HashSet<PublicKeyHash> {
        self.signing_committee
```

File: node/src/actors/chain_manager/mod.rs (hunk 1 above)
Comment (girazoki) [reply node/src/actors/chain_manager/mod.rs:1275]: definitely easier, I can modify that

File: node/src/actors/chain_manager/mod.rs [hunk 3]
```diff
@@ -2131,9 +2133,12 @@ fn current_committee_size_requirement(
     decreasing_step: u32,
     last_consolidated_checkpoint: u32,
     current_checkpoint: u32,
+    last_checkpoint_signed_by_bootstrap: u32,
 ) -> u32 {
-    // If the last consolidated superblock is 0, return the default committee size
-    if last_consolidated_checkpoint == 0 {
+    // If the last consolidated superblock or the current checkpoint is below last_checkpoint_signed_by_bootstrap, return the default committee size
+    if current_checkpoint <= last_checkpoint_signed_by_bootstrap
+        || last_consolidated_checkpoint <= last_checkpoint_signed_by_bootstrap
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mod.rs:2140]: If we assume that `last_consolidated_checkpoint <= current_checkpoint`, this can be simplified to

```suggestion
    if last_consolidated_checkpoint <= last_checkpoint_signed_by_bootstrap
```

Comment (tmpolaczyk) [reply node/src/actors/chain_manager/mod.rs:2140]: But issue #1526 shows that that assumption is not always true, for some unknown reason.

Comment (girazoki) [reply node/src/actors/chain_manager/mod.rs:2140]: I have not being able to sync with the new testnet from master so I dont think I will be able to reproduce this error. What do we think about it?

File: node/src/actors/chain_manager/mod.rs [hunk 4]
```diff
@@ -1290,6 +1290,8 @@ impl ChainManager {
                         consensus_constants.superblock_committee_decreasing_step,
                         chain_info.highest_superblock_checkpoint.checkpoint,
                         superblock_index,
+                        (chain_info.consensus_constants.collateral_age
+                            + chain_info.consensus_constants.activity_period)/superblock_period
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mod.rs:1294]: I would prefer to have this logic defined in a separate function. Something like

```rust
fn last_superblock_signed_by_bootstrap(consensus_constants: &ConsensusConstants) -> u32 {}
```

And reuse the same variable when calculating the `reputed_ars_members`.

**Review by tmpolaczyk — APPROVED** [review]

File: node/src/actors/chain_manager/mod.rs [hunk 5]
```diff
@@ -2130,9 +2135,11 @@ fn current_committee_size_requirement(
     decreasing_step: u32,
     last_consolidated_checkpoint: u32,
     current_checkpoint: u32,
+    last_checkpoint_signed_by_bootstrap: u32,
 ) -> u32 {
-    // If the last consolidated superblock is 0, return the default committee size
-    if last_consolidated_checkpoint == 0 {
+    assert!(last_consolidated_checkpoint <= current_checkpoint, "Something went wrong as the last consolidated checkpoint is bigger than our current checkpoint {} > {}", last_consolidated_checkpoint, current_checkpoint);
+    // If the last consolidated superblock or the current checkpoint is below last_checkpoint_signed_by_bootstrap, return the default committee size
+    if last_consolidated_checkpoint <= last_checkpoint_signed_by_bootstrap {
```
Comment (clbartoli) [inline node/src/actors/chain_manager/mod.rs:2142]: Maybe we could separate in two cases? Something like
- If `last_consolidated_checkpoint = current_checkpoint`, then return the default committee size
- if `last_consolidated_checkpoint < current_checkpoint`, the return the last committee size, since it should be the same as the bootstrap committee size

**Comment by girazoki** [comment]
Closed by #1557
